### ISSUE TITLE
New cop that suggests replacing complex `unless` with `if`

### ISSUE THREAD

**Issue by fatkodima** [rubocop/rubocop#8879]
I was thinking about this before, but after this thread - https://twitter.com/nateberkopec/status/1314235606705340418, I finally decided to write here to start a discussion. Turns out there are a lot of people struggling to parse non-trivial `unless` statements, and i am amongst them.

Imo, while something like `do_something unless cond` is expressive and easy to parse and understand (and even something like `do_something unless cond1 && cond2`), when conditions become more complex (having `||` or `!`) it is hard to do so.

So, I suggest adding a cop (disabled by default), that will suggest rewriting complex `unless` with `if`.
```ruby
# bad
do_something unless foo || bar
do_something unless foo || !bar
do_something unless !foo || bar || !quux
# ...

# good
do_something unless foo

do_something if !foo && !bar
do_something if !foo && bar
do_something if foo && !bar && quux

# neutral?
do_something unless foo && bar
```

**Comment by marcandre** [comment]
How should we write `do_something unless foo || bar`?

**Comment by fatkodima** [comment]
`do_something if !foo && !bar`
Updated an example.

**Comment by marcandre** [comment]
> `do_something if !foo && !bar`
> Updated an example.

I typically try to minimize the number of `!`, so I (typically) would prefer the `unless` in this case.

I could imagine a config "Prefer": 'if' / 'no negation', and maybe another one to apply only on compound conditions (as you are suggesting) or always?

Glad we agreed it shouldn't be enabled by default 😆

**Comment by natematykiewicz** [comment]
This reminds me of this Tweet. Lots of people seem to dislike `unless`. Haha.
https://twitter.com/nateberkopec/status/1314235606705340418?s=21

**Comment by dmytro-savochkin** [comment]
I was recently wondering why Rubocop haven't implement such a cop yet :) 

Personally for me conditions like this `do_something unless foo || bar` are very hard to grasp. Sometimes to understand them I have to recall De Morgan's laws to transfer them to `do_something if !foo && !bar` and only after that I can really comprehend them in my head. I would even prefer to disallow constructs like `do_something unless foo && bar` because `do_something if !foo || !bar` is easier for me.

**Comment by fatkodima** [comment]
Same for me and I'm also regularly using De Morgan's laws to understand the conditional meaning 😄 .
I completely agree with Sam Saffron from mentioned twitter thread and would prefer to have `unless` keyword banned.

**Comment by forthrin** [comment]
This is an interesting one. `unless`-es have been found harder for the brain (of the undersigned) to decipher on numerous occasions, so once again one is relieved to hear one is not the only one with something AKA nuts. Heavy lifting courtesy of LLM:

### Analysis

The underlying phenomenon is **negation-induced cognitive load**, especially **mental transformation cost**.

A simple positive condition maps directly to an action:

```ruby
do_something if ready
```

A simple `unless` is also easy:

```ruby
do_something unless busy
```

But complex `unless` conditions require the reader to:

1. Notice the negation in `unless`.
2. Parse the internal Boolean expression.
3. Invert the result mentally.
4. Apply De Morgan’s laws when `||` or `&&` is present.
5. Track any nested `!` operators.

For example:

```ruby
do_something unless foo || bar
```

must be mentally transformed into:

```ruby
do_something if !foo && !bar
```

This creates **negation scope** and **working-memory** problems. The reader must temporarily hold the original condition while constructing its inverse, increasing the chance of a polarity error.

The practical rule is:

```ruby
# Simple negation: readable
do_something unless busy

# Complex negation: prefer positive form
do_something if !foo && !bar
do_something if !foo && bar && quux
```

`unless foo && bar` is often acceptable because it contains one straightforward conjunction. `unless` becomes harder to read mainly when the condition contains `||`, nested `!`, or multiple levels of Boolean logic.

### Suggestion

Implement it as a complexity-based cop for `unless`, not as a general preference against `unless`.

```yaml
Style/ComplexUnless:
  Enabled: false
  MaxComplexity: 1
  AllowModifier: true
```

Recommended rules:

```ruby
do_something unless foo                 # allow
do_something unless foo && bar          # allow by default

do_something unless foo || bar          # flag
do_something unless !foo                # flag
do_something unless foo || !bar         # flag
do_something unless !foo || bar || quux # flag
```

Use the condition’s AST:

```ruby
def complex_unless?(condition)
  condition.each_descendant.any? do |node|
    node.or_type? || node.not_type?
  end
end
```

The cop should:

- Inspect `if` nodes where `node.unless?` is true.
- Flag conditions containing `||` or explicit `!`.
- Optionally support a configurable Boolean complexity threshold.
- Prefer modifier `unless` for simple conditions.
- Ignore ordinary `if` expressions.
- Avoid autocorrect by default unless the condition can be safely inverted.
- Handle `else` carefully, since converting `unless` to `if` must swap the branches.

Safe transformations:

```ruby
do_something unless foo || bar
```

becomes:

```ruby
do_something if !foo && !bar
```

For a full `unless`:

```ruby
unless foo || bar
  action_a
else
  action_b
end
```

the equivalent is:

```ruby
if !foo && !bar
  action_a
else
  action_b
end
```

The cop should use an AST-based De Morgan transformation rather than string manipulation, and preserve parentheses and comments. A good default policy is: **allow `unless` with one simple negation-free conjunction; flag `||`, nested negation, and deeper Boolean expressions.**

**Comment by forthrin** [comment]
Sometimes it might also be appropriate to consider whether the variables would be more expressive if inverted, eg.

`unless closed -> if open`
